sdk/program/src/message/legacy.rs

```diff
@@ -363,7 +363,8 @@ impl Message {
                 let key = self.account_keys[i];
                 sysvar::is_sysvar_id(&key) || BUILTIN_PROGRAMS_KEYS.contains(&key)
             }
-            && !(demote_program_write_locks && self.is_key_called_as_program(i))
+            && !(demote_program_write_lock_features
+                && (self.is_key_called_as_program(i) && !self.is_upgradeable_loader_present()))
```

Comment:
Logic change 1 of 2 here

Reply:
Sorry if these links suck. I did some rework to make the new cases easier to follow.